packages/core/src/browser/shell/tab-bar-decorator.ts

```diff
@@ -0,0 +1,175 @@
+import { Event, Emitter, Disposable, DisposableCollection, ContributionProvider } from '../../common';
+import { Title, Widget } from '@phosphor/widgets';
+
+export const TabBarDecorator = Symbol('TabBarDecorator');
+
+export interface TabBarDecorator {
+
+    /**
+     * The unique identifier of the tab bar decorator.
+     */
+    readonly id: string;
+
+    readonly onDidChangeDecorations: Event<(titles: Title<Widget>[]) => void>;
+
+    /**
+     * Decorate tabs by the associated editor URI.
+     * @returns A map from URI of the tab to its decoration data.
+     */
+    decorate(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data>;
+}
+
+@injectable()
+export class TabBarDecoratorService implements Disposable {
+
+    protected readonly onDidChangeDecorationsEmitter = new Emitter<void>();
+
+    readonly onDidChangeDecorations = this.onDidChangeDecorationsEmitter.event;
+
+    protected readonly toDispose = new DisposableCollection();
+
+    @inject(ContributionProvider) @named(TabBarDecorator)
+    protected readonly contributions: ContributionProvider<TabBarDecorator>;
+
+    protected decorators: ReadonlyArray<TabBarDecorator>;
+
+    @postConstruct()
+    protected init(): void {
+        this.decorators = this.contributions.getContributions();
+        this.toDispose.pushAll(this.decorators.map(decorator =>
+            decorator.onDidChangeDecorations(data =>
+                this.onDidChangeDecorationsEmitter.fire(undefined)
+            ))
+        );
+
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    getDecorations(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data[]> {
+        const changes: Map<string, TabBarDecoration.Data[]> = new Map();
+        for (const decorator of this.decorators) {
+            console.log(decorator);
+            for (const [id, data] of (decorator.decorate(titles)).entries()) {
+                if (changes.has(id)) {
+                    changes.get(id)!.push(data);
+                } else {
+                    changes.set(id, [data]);
+                }
+            }
+        }
+        return changes;
+    }
+}
+
+export namespace TabBarDecoration {
+
+    export namespace Styles {
+        export const CAPTION_HIGHLIGHT_CLASS = 'theia-caption-highlight';
+        export const CAPTION_PREFIX_CLASS = 'theia-caption-prefix';
+        export const CAPTION_SUFFIX_CLASS = 'theia-caption-suffix';
+        export const ICON_WRAPPER_CLASS = 'theia-icon-wrapper';
+        export const DECORATOR_SIZE_CLASS = 'theia-decorator-size';
+        export const TOP_RIGHT_CLASS = 'theia-top-right';
+        export const BOTTOM_RIGHT_CLASS = 'theia-bottom-right';
+        export const BOTTOM_LEFT_CLASS = 'theia-bottom-left';
+        export const TOP_LEFT_CLASS = 'theia-top-left';
+    }
+
+    /**
+     * Enumeration for the quadrant to overlay the icon on.
+     */
+    export enum IconOverlayPosition {
+        TOP_RIGHT,
```

Comment:
These enum values are undocumented while being documented in the `tree-decorator.ts`